### PR TITLE
[Kafka deprecation] Move ack level to source cluster

### PR DESCRIPTION
**Pull request by yux0**

### CODE REVIEW

File: idl/github.com/uber/cadence/replicator.thrift [hunk 1]
```diff
@@ -127,15 +127,21 @@ struct ReplicationToken {
...
+struct SyncShardStatus {
+    10: optional i64 (js.type = "Long") timestamp
+}
+
 struct ReplicationMessages {
   10: optional list<ReplicationTask> replicationTasks
   // This can be different than the last taskId in the above list, because sender can decide to skip tasks (e.g. for completed workflows).
   20: optional i64 (js.type = "Long") lastRetrievedMessageId
   30: optional bool hasMore // Hint for flow control
+  40: optional SyncShardStatus syncShardStatus
```
Comment (meiliang86) [inline idl/github.com/uber/cadence/replicator.thrift]: In shard context we have standbyClusterCurrentTime map[string]time.Time
Maybe we should rename it to remoteClusterCurrentTime, as it's not strictly standByCluster anymore.

File: service/history/replicationTaskProcessor.go [hunk 2]
```diff
@@ -200,15 +218,52 @@ Loop:
...
 		case <-p.done:
-			p.logger.Info("ReplicationTaskProcessor shutting down.")
 			return
 		}
 	}
 }
 
-func (p *ReplicationTaskProcessor) sendFetchMessageRequest() <-chan *r.ReplicationMessages {
+func (p *ReplicationTaskProcessorImpl) cleanupReplicationTaskLoop() {
+	ticker := time.NewTicker(cleanupInterval)
```
Comment (meiliang86) [inline service/history/replicationTaskProcessor.go]: Consider adding some gitter so not all the shards will do clean up at the same time?

File: service/history/replicationTaskProcessor.go [hunk 3]
```diff
@@ -240,17 +297,53 @@ func (p *ReplicationTaskProcessor) processResponse(response *r.ReplicationMessag
...
-
 	scope := p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster))
 	scope.UpdateGauge(metrics.LastRetrievedMessageID, float64(p.lastRetrievedMessageID))
 	p.noTaskRetrier.Reset()
 }
 
-func (p *ReplicationTaskProcessor) processSingleTask(replicationTask *r.ReplicationTask) error {
+func (p *ReplicationTaskProcessorImpl) syncShardStatusLoop() {
+
+	ticker := time.NewTicker(p.config.ShardSyncMinInterval())
```
Comment (meiliang86) [inline service/history/replicationTaskProcessor.go]: Same here. Consider adding some gitter.

File: service/history/replicatorQueueProcessor.go [hunk 4]
```diff
@@ -471,6 +477,13 @@ func (p *replicatorQueueProcessorImpl) getTasks(
 		time.Duration(len(replicationTasks)),
 	)
 
+	if err := p.shard.UpdateClusterReplicationLevel(
+		sourceCluster,
+		lastReadTaskID,
```
Comment (meiliang86) [inline service/history/replicatorQueueProcessor.go]: lastReadTaskID cannot be updated in this way. Imagine that this batch is totally lost before sending. Then poller will never see it again.
lastReadTaskID should be whatever value passed from the poller.

File: service/history/shardController_test.go [hunk 5]
```diff
@@ -23,11 +23,12 @@ package history
 import (
 	"errors"
 	"fmt"
-	"github.com/uber/cadence/common/service/dynamicconfig"
 	"sync"
 	"testing"
 	"time"
 
+	"github.com/uber/cadence/common/service/dynamicconfig"
+
```
Comment (meiliang86) [inline service/history/shardController_test.go]: remove extra line.

File: common/persistence/cassandra/cassandraPersistence.go [hunk 6]
```diff
@@ -734,6 +734,15 @@ workflow_state = ? ` +
 		`and task_id > ? ` +
 		`and task_id <= ?`
 
+	templateEndRangeCompleteTransferTaskQuery = `DELETE FROM executions ` +
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/cassandra/cassandraPersistence.go]: Updated the name

File: common/persistence/sql/sqlplugin/mysql/execution.go [hunk 7]
```diff
@@ -306,13 +307,16 @@ func (mdb *db) InsertIntoReplicationTasks(rows []sqlplugin.ReplicationTasksRow)
 // SelectFromReplicationTasks reads one or more rows from replication_tasks table
 func (mdb *db) SelectFromReplicationTasks(filter *sqlplugin.ReplicationTasksFilter) ([]sqlplugin.ReplicationTasksRow, error) {
 	var rows []sqlplugin.ReplicationTasksRow
-	err := mdb.conn.Select(&rows, getReplicationTasksQuery, filter.ShardID, filter.MinTaskID, filter.MaxTaskID, filter.PageSize)
+	err := mdb.conn.Select(&rows, getReplicationTasksQuery, filter.ShardID, *filter.MinTaskID, *filter.MaxTaskID, *filter.PageSize)
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/sql/sqlplugin/mysql/execution.go]: Removed the pointers

File: service/history/historyEngine.go [hunk 8]
```diff
@@ -95,7 +95,7 @@ type (
 		ReplicateEventsV2(ctx ctx.Context, request *h.ReplicateEventsV2Request) error
 		SyncShardStatus(ctx ctx.Context, request *h.SyncShardStatusRequest) error
 		SyncActivity(ctx ctx.Context, request *h.SyncActivityRequest) error
-		GetReplicationMessages(ctx ctx.Context, taskID int64) (*r.ReplicationMessages, error)
+		GetReplicationMessages(ctx ctx.Context, sourceCluster string, lastReadMessageID int64) (*r.ReplicationMessages, error)
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/historyEngine.go]: Bad name for this one. The sourceCluster is the remote cluster. I updated the name to be remote cluster

File: idl/github.com/uber/cadence/replicator.thrift (hunk 1 above)
Comment (yux0) [reply idl/github.com/uber/cadence/replicator.thrift]: updated the field in shard context

File: service/history/replicationTaskProcessor.go (hunk 2 above)
Comment (yux0) [reply service/history/replicationTaskProcessor.go]: added jitter

File: service/history/replicationTaskProcessor.go (hunk 3 above)
Comment (yux0) [reply service/history/replicationTaskProcessor.go]: added jitter

File: service/history/replicatorQueueProcessor.go (hunk 4 above)
Comment (yux0) [reply service/history/replicatorQueueProcessor.go]: Synced offline. the lastReadTaskID does not update and it is from the remote clusters

File: service/history/shardController_test.go (hunk 5 above)
Comment (yux0) [reply service/history/shardController_test.go]: Sorted imports

File: service/history/historyEngine.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (meiliang86) [reply service/history/historyEngine.go]: Maybe call it pollingCluster

**Pull request merged**